### PR TITLE
Account for inaccurate offsets in getXrefData()

### LINKED ISSUE #673: Undefined array key 1,crash on parsing
- PHP Versiov 8.3: 
 - PDFParser Version 2.8: 

### PDF input

[Cracking-the-Coding-Interview-6th-Edition-189-Programming-Questions-and-Solutions.pdf](https://github.com/smalot/pdfparser/files/14177782/Cracking-the-Coding-Interview-6th-Edition-189-Programming-Questions-and-Solutions.pdf)

### Expected output & actual output

It crashes in the RawDataPraser, line 890 

`elseif ($startxrefPreg) {
            // startxref found
            $startxref = $matches[1][0];
        }`
$matches is empty array.
### Code

Just the usual stuff.

### PR DESCRIPTION
**Pull request by GreyWyvern**
# Type of pull request

* [X] Bug fix (involves code and configuration changes)

# About

Normally offset pointers to `xref` keywords in a PDF document are exact to the byte. However, in some cases the pointer may point to some whitespace *before* the `xref` keyword. Adobe Acrobat takes these 'errors' in stride, displaying the document anyway, and so should PdfParser.

Clean up the getXrefData() function in **RawDataParser.php**. It now only needs to do one `preg_match_all()` and pushes the caret past whitespace when looking for `xref` keywords.

Use existing **Issue557.pdf** to create a new file: **Issue673.pdf** where the last `/Prev 13486` command has been decremented to `/Prev 13485`. Trying to parse this file would cause an Exception without this fix. Resolves #673.

# Checklist for code / configuration changes

*In case you changed the code/configuration, please read each of the following checkboxes as they contain valuable information:*

* [X] Please add at least **one test case** (unit test, system test, ...) to demonstrate that the change is working. If existing code was changed, your tests cover these code parts as well.
* [X] Please run **PHP-CS-Fixer** before committing, to confirm with our coding styles. See https://github.com/smalot/pdfparser/blob/master/.php-cs-fixer.php for more information about our coding styles.
* [X] In case you **fix an existing issue**, please do one of the following:
  * [X] Write in this text something like `fixes #1234` to outline that you are providing a fix for the issue `#1234`.

### CODE REVIEW

File: src/Smalot/PdfParser/RawData/RawDataParser.php [hunk 1]
```diff
@@ -864,39 +864,39 @@ private function getHeaderValue(?array $headerDic, string $key, string $type, $d
...
+        // Find all startxref tables from this $offset forward
+        $startxrefPreg = preg_match_all(
+            '/(?<=[\r\n])startxref[\s]*[\r\n]+([0-9]+)[\s]*[\r\n]+%%EOF/i',
             $pdfData,
-            $matches,
-            \PREG_OFFSET_CAPTURE,
+            $startxrefMatches,
+            \PREG_SET_ORDER,
             $offset
         );
```
Comment (k00ni) [inline src/Smalot/PdfParser/RawData/RawDataParser.php:883]: Using `$startxrefMatches` as third parameter and to store result of `preg_match_all` looks weird. It seems to work, but I am wondering which one of these two is stored. Based on your code you assume its the latter. But you also use it as it contains the matches (e.g. line 890).

I would like to use two separate variables here to improve code readability.

File: src/Smalot/PdfParser/RawData/RawDataParser.php [hunk 2]
```diff
@@ -864,39 +864,39 @@ private function getHeaderValue(?array $headerDic, string $key, string $type, $d
...
             // Cross-Reference Stream object
-            $startxref = $offset;
-        } elseif ($startxrefPreg) {
-            // startxref found
-            $startxref = $matches[1][0];
+            $startxref = $bumpOffset;
         } else {
-            throw new \Exception('Unable to find startxref');
+            // Use the next startxref from this $offset
+            $startxref = (int) $startxrefMatches[0][1];
```
Comment (k00ni) [inline src/Smalot/PdfParser/RawData/RawDataParser.php:899]: Can we be sure that `$startxrefMatches[0][1]` is always defined? If not, we could run into an index not defined error.

**Review by k00ni — CHANGES_REQUESTED** [review]
Thank you @GreyWyvern, good work! I have just a few remarks/questions.

File: src/Smalot/PdfParser/RawData/RawDataParser.php (hunk 2 above)
Comment (GreyWyvern) [reply src/Smalot/PdfParser/RawData/RawDataParser.php:899]: On line 885, there is a non-strict check of the `preg_match_all()` return value `0 == $startxrefPreg` which will catch if there were no matches, or if there was a failure `false`. So to arrive at the final `else` here, `$startxrefMatches` MUST have at least one match in it.

File: src/Smalot/PdfParser/RawData/RawDataParser.php (hunk 1 above)
Comment (GreyWyvern) [reply src/Smalot/PdfParser/RawData/RawDataParser.php:883]: `$startxrefPreg` is the result of the `preg_match_all()` and `$startxrefMatches` stores the associated matches. So they are already two different variables? Unless I'm reading your comment incorrectly...

Comment (k00ni) [reply src/Smalot/PdfParser/RawData/RawDataParser.php:883]: My bad, all good.

**Review by k00ni — APPROVED** [review]
Sorry for the delay. I merged in a recent change to get rid of the coding style issue.

**Pull request merged**